cmd/agent/api/clc_runner_server.go

```diff
@@ -0,0 +1,114 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+/*
+Package api implements the agent IPC api. Using HTTP
+calls, it's possible to communicate with the agent,
+sending commands and receiving infos.
+*/
+package api
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	stdLog "log"
+	"net"
+	"net/http"
+	"time"
+
+	v1 "github.com/DataDog/datadog-agent/cmd/agent/api/v1"
+	"github.com/DataDog/datadog-agent/pkg/api/security"
+	"github.com/DataDog/datadog-agent/pkg/api/util"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/gorilla/mux"
+)
+
+var (
+	clcListener net.Listener
+)
+
+// StartCLCRunnerServer creates the router and starts the HTTP server
+func StartCLCRunnerServer() error {
+	// create the root HTTP router
+	r := mux.NewRouter()
+
+	// IPC REST API server
+	v1.SetupHandlers(r.PathPrefix("/api/v1").Subrouter())
+
+	// Validate token for every request
+	r.Use(validateCLCRunnerToken)
+
+	// get the transport we're going to use under HTTP
+	var err error
+	clcListener, err = getCLCRunnerListener()
+	if err != nil {
+		return fmt.Errorf("unable to create the clc runner api server: %v", err)
+	}
+
+	// CLC Runner token
+	// Use the Cluster Agent token
+	err = util.SetDCAAuthToken()
+	if err != nil {
+		return err
+	}
+
+	hosts := []string{"127.0.0.1", "localhost", config.Datadog.GetString("clc_runner_host")}
+	_, rootCertPEM, rootKey, err := security.GenerateRootCert(hosts, 2048)
+	if err != nil {
+		return fmt.Errorf("unable to start TLS server: %v", err)
+	}
+
+	// PEM encode the private key
+	rootKeyPEM := pem.EncodeToMemory(&pem.Block{
+		Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(rootKey),
+	})
+
+	// Create a TLS cert using the private key and certificate
+	rootTLSCert, err := tls.X509KeyPair(rootCertPEM, rootKeyPEM)
+	if err != nil {
+		return fmt.Errorf("invalid key pair: %v", err)
+	}
+
+	tlsConfig := tls.Config{
+		Certificates: []tls.Certificate{rootTLSCert},
+	}
+
+	srv := &http.Server{
+		Handler: r,
+		ErrorLog: stdLog.New(&config.ErrorLogWriter{
+			AdditionalDepth: 4, // Use a stack depth of 4 on top of the default one to get a relevant filename in the stdlib
+		}, "Error from the clc runner http API server: ", 0), // log errors to seelog,
+		TLSConfig:    &tlsConfig,
+		WriteTimeout: config.Datadog.GetDuration("server_timeout") * time.Second,
```

Comment:
I think it's necessary to set a timeout, we don't want to overload the Agent with "long" open http connections.
For the cluster agent, we're relying on the DCA client timeout which is 2 sec (not configurable) (and which is not the best way to do so, I think)

Totally agree on the second point, although I wanted to keep the new server's code as similar as possible to the internal cmd http server to make it easier to maintain both servers.

@DataDog/agent-core we would appreciate your suggestion

Reply:
My feeling is there should be a connection timeout if only for graceful behavior on the client side in case of the server agent misbehaving. It would also have the positive effect of preventing connection build-up on the server-side. 

Making it configurable would be important, I totally agree no hard-coded timeout is good enough it's always hard to anticipate what a sane value is and it may vary from environment to environment. 

Although I do agree that keeping configuration for both the private and public facing API endpoints as similar as possible, they do have different constraints and I guess that's a legitimate reason for the two to diverge if necessary. From a networking perspective the number of things that can go wrong accessing a remote endpoint are far beyond your typical considerations for a local connection. 

Also, the option should probably be called `client_timeout`, no? As this would be a timeout enforced on the client side - probably not too important, especially if we're already using it like this.